File: app/src/commands/military.ts
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

import { invoke } from "@tauri-apps/api/core";
import { CoordImpl } from "@/core/model/continent/coord";
import type { ContinentKey } from "@/types/core/continent";
import type {
  ArmyId,
  CancelManeuverRequest,
  GetArmiesRequest,
  GetArmiesResponse,
  GetArmyOwnerRequest,
  GetArmyOwnerResponse,
  GetArmyRequest,
  GetArmyResponse,
  GetIdleArmiesAtRequest,
  GetIdleArmiesAtResponse,
  GetIdleArmiesCoordsRequest,
  GetIdleArmiesCoordsResponse,
  GetManeuverRequest,
  GetManeuverResponse,
  ManeuverId,
  ManeuverRequest,
  RequestManeuverRequest,
  RequestManeuverResponse,
} from "@tsukilabs/nil-bindings";

export async function cancelManeuver(id: ManeuverId) {
  const req: CancelManeuverRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  await invoke("cancel_maneuver", { req });
}

export async function getArmies() {
  const req: GetArmiesRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetArmiesResponse>("get_armies", { req });
}

export async function getArmy(id: ArmyId) {
  const req: GetArmyRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<GetArmyResponse>("get_army", { req });
}

export async function getArmyOwner(id: ArmyId) {
  const req: GetArmyOwnerRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<GetArmyOwnerResponse>("get_army_owner", { req });
}

export async function getIdleArmiesAt(coord: ContinentKey) {
  coord = CoordImpl.fromContinentKey(coord);
  const req: GetIdleArmiesAtRequest = {
    world: NIL.world.getIdStrict(),
    coord,
  };

  return invoke<GetIdleArmiesAtResponse>("get_idle_armies_at", { req });
}

export async function getIdleArmiesCoords() {
  const req: GetIdleArmiesCoordsRequest = {
    world: NIL.world.getIdStrict(),
  };

  return invoke<GetIdleArmiesCoordsResponse>("get_idle_armies_coords", { req });
}

export async function getManeuver(id: ManeuverId) {
  const req: GetManeuverRequest = {
    world: NIL.world.getIdStrict(),
    id,
  };

  return invoke<GetManeuverResponse>("get_maneuver", { req });
}

export async function requestManeuver(request: ManeuverRequest) {
  const req: RequestManeuverRequest = {
    world: NIL.world.getIdStrict(),
    request,
  };

  return invoke<RequestManeuverResponse>("request_maneuver", { req });
}

File: crates/nil-server/src/router/military.rs
// Copyright (C) Call of Nil contributors
// SPDX-License-Identifier: AGPL-3.0-only

use crate::app::App;
use crate::error::CoreError;
use crate::middleware::authorization::CurrentPlayer;
use crate::res;
use crate::response::{EitherExt, from_err};
use axum::extract::{Extension, Json, State};
use axum::response::Response;
use itertools::Itertools;
use nil_core::military::army::Army;
use nil_core::ruler::Ruler;
use nil_payload::request::military::*;
use nil_payload::response::military::*;
use tap::Pipe;

pub async fn cancel_maneuver(
  State(app): State<App>,
  Extension(player): Extension<CurrentPlayer>,
  Json(req): Json<CancelManeuverRequest>,
) -> Response {
  match app.get(req.world) {
    Ok(world) => {
      let result = try {
        let mut world = world.write().await;
        if world.is_maneuver_army_owned_by(req.id, player)? {
          world.cancel_maneuver(req.id)?;
        } else {
          return res!(FORBIDDEN);
        }
      };

      result
        .map(|()| res!(OK))
        .unwrap_or_else(from_err)
    }
    Err(err) => from_err(err),
  }
}

pub async fn get_armies(
  State(app): State<App>,
  Extension(player): Extension<CurrentPlayer>,
  Json(req): Json<GetArmiesRequest>,
) -> Response {
  app
    .world(req.world, |world| {
      let k_size = world.continent().size();
      world
        .military()
        .indexed_armies_of(player)
        .into_iter()
        .map(|(index, army)| Ok((index.to_coord(k_size)?, army.clone())))
        .try_collect::<_, _, CoreError>()
    })
    .await
    .try_map_left(|armies| res!(OK, GetArmiesResponse(armies)))
    .into_inner()
}

pub async fn get_army(
  State(app): State<App>,
  Extension(player): Extension<CurrentPlayer>,
  Json(req): Json<GetArmyRequest>,
) -> Response {
  match app.get(req.world) {
    Ok(world) => {
      let result = try {
        let army = world
          .read()
          .await
          .military()
          .army(req.id)
          .cloned()?;

        if player != *army.owner() {
          return res!(FORBIDDEN);
        }

        army
      };

      result
        .map(|army| res!(OK, GetArmyResponse(army)))
        .unwrap_or_else(from_err)
    }
    Err(err) => from_err(err),
  }
}

pub async fn get_army_owner(
  State(app): State<App>,
  Json(req): Json<GetArmyOwnerRequest>,
) -> Response {
  app
    .military(req.world, |military| {
      military
        .army(req.id)
        .map(Army::owner)
        .cloned()
    })
    .await
    .try_map_left(|owner| res!(OK, GetArmyOwnerResponse(owner)))
    .into_inner()
}

pub async fn get_idle_armies_at(
  State(app): State<App>,
  Extension(player): Extension<CurrentPlayer>,
  Json(req): Json<GetIdleArmiesAtRequest>,
) -> Response {
  app
    .world(req.world, |world| {
      let player: Ruler = player.into();
      let is_own_city = world
        .continent()
        .city(req.coord)?
        .is_owned_by(player.clone());

      world
        .military()
        .idle_armies_at(req.coord)
        .filter(|army| is_own_city || army.is_owned_by(&player))
        .cloned()
        .collect_vec()
        .pipe(Ok::<_, CoreError>)
    })
    .await
    .try_map_left(|armies| res!(OK, GetIdleArmiesAtResponse(armies)))
    .into_inner()
}

pub async fn get_idle_armies_coords(
  State(app): State<App>,
  Extension(player): Extension<CurrentPlayer>,
  Json(req): Json<GetIdleArmiesCoordsRequest>,
) -> Response {
  app
    .world(req.world, |world| {
      let k_size = world.continent().size();
      world
        .military()
        .indexed_armies_of(player)
        .into_iter()
        .filter(|(_, army)| army.is_idle())
        .map(|(index, _)| index.to_coord(k_size))
        .try_collect()
    })
    .await
    .try_map_left(|armies| res!(OK, GetIdleArmiesCoordsResponse(armies)))
    .into_inner()
}

pub async fn get_maneuver(State(app): State<App>, Json(req): Json<GetManeuverRequest>) -> Response {
  app
    .military(req.world, |military| {
      match military.maneuver(req.id) {
        Ok(maneuver) => Ok(Some(maneuver.clone())),
        Err(CoreError::ManeuverNotFound(_)) => Ok(None),
        Err(err) => Err(err),
      }
    })
    .await
    .try_map_left(|maneuver| res!(OK, GetManeuverResponse(maneuver)))
    .into_inner()
}

pub async fn request_maneuver(
  State(app): State<App>,
  Extension(player): Extension<CurrentPlayer>,
  Json(mut req): Json<RequestManeuverRequest>,
) -> Response {
  req.request.ruler = Ruler::from(&player.0);
  app
    .world_blocking_mut(req.world, move |world| {
      if world.round().is_waiting_player(&player.0) {
        world.request_maneuver(req.request)
      } else {
        Err(CoreError::NotWaitingPlayer(player.0))
      }
    })
    .await
    .try_map_left(|id| res!(OK, RequestManeuverResponse(id)))
    .into_inner()
}
